alacritty_terminal/src/ansi.rs

```diff
@@ -25,60 +25,88 @@ use vte;
 use crate::term::color::Rgb;
 
 // Parse color arguments
-//
-// Expect that color argument looks like "rgb:xx/xx/xx" or "#xxxxxx"
-fn parse_rgb_color(color: &[u8]) -> Option<Rgb> {
-    let mut iter = color.iter();
 
-    macro_rules! next {
-        () => {
-            iter.next().map(|v| *v as char)
-        };
+// Parse a color argument in the form "#xxxxxx"
+// The use of this format is no longer encouraged by xparsecolor
+fn parse_legacy_rgb_color(color: &[u8]) -> Option<Rgb> {
+    if color.len() < 7 || color[0] != b'#' {
+        return None;
     }
 
+    let mut iter = color.iter().map(|v| (*v as char));
+    iter.next();
+
     macro_rules! parse_hex {
         () => {{
-            let mut digit: u8 = 0;
-            let next = next!().and_then(|v| v.to_digit(16));
-            if let Some(value) = next {
-                digit = value as u8;
-            }
+            let mut hex: u8 = match iter.next().and_then(|v| v.to_digit(16)) {
+                Some(digit) => digit as u8,
+                _ => return None,
+            };
+            hex <<= 4;
+            hex += match iter.next().and_then(|v| v.to_digit(16)) {
+                Some(digit) => digit as u8,
+                _ => return None,
+            };
+            hex
+        }};
+    }
+
+    let r = parse_hex!();
+    let g = parse_hex!();
+    let b = parse_hex!();
 
-            let next = next!().and_then(|v| v.to_digit(16));
-            if let Some(value) = next {
-                digit <<= 4;
-                digit += value as u8;
+    Some(Rgb { r, g, b })
```

Comment:
I know that you've probably just extended what already existed in this method, but calling a macro that takes things out of the current scope and even mutates it is *really* bad style. So I'd like to take the chance to improve the style and do it right here.

This is what I'd propose for this method to look like, including the previous change I've mentioned:

```rust
fn parse_legacy_rgb_color(color: &[u8]) -> Option<Rgb> {
    if color.len() < 7 || color[0] != b'#' {
        return None;
    }

    let mut iter = color.iter().map(|v| (*v as char)).skip(1).skip(1);

    let mut colors = [0; 3];
    for i in 0..3 {
        colors[i] = iter.next().and_then(|v| v.to_digit(16))? as u8;

        colors[i] <<= 4;

        colors[i] += iter.next().and_then(|v| v.to_digit(16))? as u8;
    }

    Some(Rgb { r: colors[0], g: colors[1], b: colors[2] })
}
```